### PR TITLE
Improve the `DescribeIndex` RPC about the state and row num

### PR DESCRIPTION
**Pull request by SimFG**
issue: #19445
Signed-off-by: SimFG <bang.fu@zilliz.com>

### CODE REVIEW

File: internal/indexcoord/index_coord.go [hunk 1]
```diff
@@ -582,31 +588,37 @@ func (i *IndexCoord) completeIndexInfo(ctx context.Context, indexInfo *indexpb.I
...
+	switch {
+	case indexStateCnt.Failed > 0:
+		indexInfo.State = commonpb.IndexState_Failed
+		indexInfo.IndexStateFailReason = indexStateCnt.FailReason
+	case indexStateCnt.Finished == allCnt:
+		indexInfo.State = commonpb.IndexState_Finished
+	default:
+		indexInfo.State = commonpb.IndexState_InProgress
+		indexInfo.IndexedRows = i.metaTable.GetIndexBuildProgress(indexID, createTs)
+		totalRow, err := calculateTotalRow()
```
Comment (xiaocai2333) [inline internal/indexcoord/index_coord.go:612]: modify this function, get index rows by filter segments.

File: internal/indexcoord/index_coord.go [hunk 2]
```diff
@@ -550,30 +550,36 @@ func (i *IndexCoord) GetSegmentIndexState(ctx context.Context, req *indexpb.GetS
...
+			CollectionID: collectionID,
+			PartitionID:  -1,
+		})
+		if err != nil {
+			return totalRows, err
+		}
 
-	for _, seg := range resp.Infos {
-		totalRows += seg.NumOfRows
+		resp, err := i.dataCoordClient.GetSegmentInfo(ctx, &datapb.GetSegmentInfoRequest{
```
[the remark this answers is not in the document]
Comment (SimFG) [reply internal/indexcoord/index_coord.go:570]: the resp only includes the id, the filter logic in the next

File: internal/indexcoord/index_coord.go [hunk 3]
```diff
@@ -550,30 +550,36 @@ func (i *IndexCoord) GetSegmentIndexState(ctx context.Context, req *indexpb.GetS
...
+			SegmentIDs: flushSegments.Segments,
+		})
+		if err != nil {
+			return totalRows, err
+		}
+
+		for _, seg := range resp.Infos {
+			if seg.State == commonpb.SegmentState_Flushed {
```
Comment (SimFG) [inline internal/indexcoord/index_coord.go:578]: filter the dropped segments

**Comment by SimFG** [comment]
handle the conflict

**Comment by SimFG** [comment]
rerun ut

**Comment by SimFG** [comment]
rerun ut

**Pull request merged**